libbeat/registry/backend/memlog/typeconv.go

```diff
@@ -0,0 +1,88 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memlog
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/go-structform/gotype"
+)
+
+type typeConv struct {
+	fold   *gotype.Iterator
+	unfold *gotype.Unfolder
+}
+
+type valueDecoder struct {
+	tx    memTxParent
+	value common.MapStr
+}
+
+var typeConvPool = sync.Pool{
+	New: func() interface{} {
+		t := &typeConv{}
+		t.init()
+		return t
+	},
+}
+
+func newTypeConv() *typeConv {
+	tc := typeConvPool.Get().(*typeConv)
+	tc.init()
+	return tc
+}
+
+func (t *typeConv) release() {
+	typeConvPool.Put(t)
+}
+
+func (t *typeConv) init() {
+	unfold, _ := gotype.NewUnfolder(nil)
+	fold, err := gotype.NewIterator(unfold)
+	if err != nil {
+		panic(err)
```

Comment:
is it possible to avoid panic here? 
this does not return error but maybe it should

Reply:
A panic here would indicate a programmers error one can never recover from. We should not hide failing invariantes behind error values which can be ignored (e.g. logged, but system can never recover).

Execution of all methods are covered by the unit tests, but indirectly. Will introduce tests for typeconv.go, so it's ensure that code paths that can panic will do so during tests.